### DISCUSSION TITLE
When Actions fail

### DISCUSSION THREAD

**Discussion by teledactyl** [microsoft/teams-sdk#1578] [Q&A] [answered]
I'm really impressed with the way the Bot can infer parameters for my Actions: I'm using it to manipulate a 3D scene in the Teams Stage. In an industrial simulation I can say "Hide the robot" and it will change the visibility of the robot model by triggering the 'Hide' command with the 'part' parameter set to "robot". The response is something like: "I've hidden the robot".
But if I ask it to "Hide the cauliflower" then it will happily say it has hidden the cauliflower even though an `if` in my command implementation directed it to fail with the response 'part not found'.

Exactly what is the return `string` used for in the `handler` function?
https://learn.microsoft.com/en-us/javascript/api/@microsoft/teams-ai/ai?view=msteams-client-js-latest#@microsoft-teams-ai-ai-action

My action code is more or less:
```
app.ai.action('HidePart', async (context: TurnContext, state: ApplicationTurnState, parameters: HideShowParameters) => {
  const controller = sdcApi.controller?.scene;
  const node1 = findSceneNodeFromPartInstance(parameters.part, controller);
  if (node1) {
    controller?.updateVisibility(node1, false);
    return `the part is now invisible`;
  } else {
    return `part not found`;
  }
});
```
Do I need to use a `PromptResponseValidator` or the `completePrompt` function in the `ActionPlanner`? Is there an example? Do I need to be more persuasive with my prompts? 
Many thanks

Steve

**Comment by singhk97** [comment, marked as answer]
If you're using the sequence augmentation the planner will generate the plan in it's entirety. And sometimes it will include a command to say to the user "I've hidden the cauliflower" even before the "HidePart" executes. Here are a few options to resolve this:

1. What you can do is use the [`state.temp.actionOutputs`](https://github.com/microsoft/teams-ai/blob/720bf1c3c54b6366a3d66f52df32f5821c2e07cd/js/packages/teams-ai/src/TurnState.ts#L71) property to access the return value of any action and within the `PredictedSayCommand` (i.e that command that sends the message back to the user) and conditionally send the appropraite response to the user.

2. An alternative is to send a message back to the user within `HidePart` action itself and terminating the plan by returning `AI.StopActionName`.

**Reply by singhk97** [reply]
I also recommend looking into the ActionPlanner [docs](https://github.com/microsoft/teams-ai/blob/main/getting-started/CONCEPTS/ACTION-PLANNER.md).

**Reply by teledactyl** [reply]
The `AI.StopActionName` is a really simple solution. Many thanks.
I was reading in the docs about the differences between 'sequence' and 'monologue' augmentations as another potential solution. 
I'm tempted to ask if this can be changed at runtime (e.g. a setting to 'adopt a chain of thought' reasoning).

**Comment by corinagum** [comment]
Closing as this is related to v1.
